--- a/include/rabitqlib/quantization/rabitq_impl.hpp
+++ b/include/rabitqlib/quantization/rabitq_impl.hpp
@@ -1,6 +1,8 @@
 #pragma once
 
+#ifdef _OPENMP
 #include <omp.h>
+#endif
 
 #include <cmath>
 #include <cstddef>
--- a/include/rabitqlib/utils/space.hpp
+++ b/include/rabitqlib/utils/space.hpp
@@ -1,6 +1,8 @@
 #pragma once
 
+#ifdef _OPENMP
 #include <omp.h>
+#endif
 
 #include <array>
 #include <cassert>
@@ -154,12 +156,20 @@
 inline std::vector<T> compute_centroid(
     const T* data, size_t num_points, size_t dim, size_t num_threads
 ) {
+#ifdef _OPENMP
     omp_set_num_threads(static_cast<int>(num_threads));
+#else
+    num_threads = 1;
+#endif
     std::vector<std::vector<T>> all_results(num_threads, std::vector<T>(dim, 0));
 
 #pragma omp parallel for schedule(dynamic)
     for (size_t i = 0; i < num_points; ++i) {
+#ifdef _OPENMP
         auto tid = omp_get_thread_num();
+#else
+        const int tid = 0;
+#endif
         std::vector<T>& cur_results = all_results[tid];
         const T* cur_data = data + (dim * i);
         for (size_t k = 0; k < dim; ++k) {
@@ -195,6 +205,10 @@
 
 #pragma omp parallel for schedule(dynamic)
     for (size_t i = 0; i < num_points; ++i) {
+#ifdef _OPENMP
         auto tid = omp_get_thread_num();
+#else
+        const int tid = 0;
+#endif
         AnnCandidate<T, PID>& cur_entry = best_entries[tid];
         const T* cur_data = data + (dim * i);
